src/LazilyInitializedFields.jl

```diff
@@ -79,6 +79,50 @@ Base.show(io::IO, u::Uninitialized) = print(io, "uninit")
 # islazyfield(::Type{Foo}, s::Symbol) = s === :a || s === :b
 function islazyfield end
 
+# For mutable types without type parameters we create a Box
+# for the lazy fields instead of making the whole struct mutable
+# https://github.com/JuliaLang/julia/issues/35053
+mutable struct Box{T}
+    x::T
+end
+Base.getindex(b::Box) = b.x
+Base.setindex!(b::Box, v) = b.x = v
+Base.convert(::Type{Box{T}}, x) where {T} = Box{T}(x)
+Base.show(io::IO, b::Box) = show(io, b.x)
+
+
+# This is a replication of the Nothing and Missing conversion functionality from Base.
+if isdefined(Base, :typesplit)
+    nonuninittype(::Type{T}) where {T} = Base.typesplit(T, Uninitialized)
+else
+    nonuninittype(::Type{T}) where {T} = Core.Compiler.typesubtract(T, Uninitialized)
+end
+promote_rule(T::Type{Uninitialized}, S::Type) = Union{S, Uninitialized}
+function promote_rule(T::Type{>:Uninitialized}, S::Type)
+    R = nonuninittype(T)
+    R >: T && return Any
+    T = R
+    R = promote_type(T, S)
+    return Union{R, Uninitialized}
+end
+
+function nonuninittype_checked(T::Type)
+    R = nonuninittype(T)
+    R >: T && error("could not compute non-uninit type")
+    return R
+end
+
+# These are awful
+Base.convert(::Type{T}, x::T) where {T>:Uninitialized} = x
+Base.convert(::Type{T}, x) where {T>:Uninitialized} = convert(nonuninittype_checked(T), x)
```

Comment:
Out of curiosity, I'm guessing you didn't use
```suggestion
Base.convert(::Type{Uninitialized}, x) = convert(nonuninittype_checked(T), x)
```
because of method ambiguities?

At that point, why don't you just write the constructor for parametric types, like you suggested?

Reply:
Because I don't know how to write it in a way that will work properly.

Reply:
`Union` and `Nothing` has these type of wide conversion methods and in case we have a 

```
@lazy struct Foo
    @lazy a::Union{Nothing, Float32}
end
Foo(1)
```

these conversions seem to be needed